doc/ug/rust_for_c.md

```diff
@@ -0,0 +1,2305 @@
+The compiler will also ensure that every value will get matched with one of the match arms, either because every case is covered (e.g., every enum variant is present) or an _irrefutable pattern_ is present (i.e., a patte which matches all values). `_`, `foo`, `(a, _)`, and `MyStruct { a, .. }` are all examples of irrefutable patterns.
+
+If the value being matched on is a reference of some kind, bound names will be references, too.
+
+For example:
+```rust
+match &my_struct {
+  MyStruct { a, .. } => {
+    // Here, `a` is a `&i32`, which is a reference to the `a` field in my_struct.
+  },
+}
+```
+This feature is sometimes called _match ergonomics_, since before it was added, explicit dereference and special `ref` pattern qualifiers had to be added to matches on references.
+
+In addition, `match` statements support two additional features on top of the pattern syntax discussed above:
+*   _Multi-match arms_ can allow a match arm to match on one of multiple patterns: `a | b | c => ...,`.
+    If any of the patterns match, the arm executes.
+*   _Match guards_ give you a shorthand for conditioning an arm on some expression: `Some(foo) if foo.has_condition() => ...,`.
+
+Also, the standard library provides the `matches!()` macro as a shorthand for the following common match expression:
+```rust
+match expr {
+  <some_complex_match_arm> => true,
+  _ => false,
+}
+// ... can be replaced with ...
+matches!(expr, some_complex_match_arm)
+```
+`matches!` supports multi-match and match guards as well.
+
+Irrefutable patterns can be used with normal variable declaration.
+The syntax `let x = ...;` actually uses a pattern: `x` is a pattern.
+Other irrefutable patterns can be used there:
+```rust
+let (a, b) = ...;  // Destructure a tuple, rather than using clunky `.0` and `.1` field names.
+let Foo { foo, bar, baz } = ...;  // Destructure a struct, to access its fields directly.
+let 42 = ...;  // Not allowed: `42` is not an irrefutable pattern.
+```
+
+Special variants of `if` and `while` exist to take advantage of patterns, too:
+```rust
+if let Some(x) = my_option {
+  // If the pattern succeeds, the body will be executed, and `x` will be bound
+  // to the value inside the Option.
+} else {
+  // Else block is optional.
+}
+
+while let Some(x) = some_func() {
+  // Loop terminates once the pattern match fails. Again, `x` is bound
+  // to the value inside the Option.
+}
+```
+Unlike normal `let` statements, `if let` and `while let` expressions are meant to be used with refutable patterns.
+
+In general, almost every place where a value is bound can be an irrefutable pattern, such as function parameters and `for` loop variables:
+```rust
+fn get_first((x, _): (u32, u32)) -> u32 { x }
+
+for (k, v) in my_key_values { .. }
+```
+
+### Traits {#traits}
+
+Traits are Rust's core code-reuse abstraction.
+Rust traits are like interfaces in other languages: a list of methods that a type must implement.
+Traits themselves, however, are _not _types.
```

Comment:
```suggestion
Traits themselves, however, are _not_ types.
```

Reply:
This one we can blame on the conversion tool I used. =P